### PR TITLE
Prompt

### PR DESCRIPTION
**Pull request by harou24**

### CODE REVIEW

**Comment by harou24** [comment]
@s-t-a-n  I got this error:
Fetching submodules
  /usr/bin/git submodule sync --recursive
  /usr/bin/git -c protocol.version=2 submodule update --init --force --depth=1 --recursive
  Error: fatal: No url found for submodule path 'extern/get_next_line_lib' in .gitmodules
  Error: The process '/usr/bin/git' failed with exit code 128

**Review by s-t-a-n — CHANGES_REQUESTED** [review]
You have added a broken submodule in a commit called 'test is working'. That's why it's failing.

File: includes/prompt.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#ifndef PROMPT_H
+# define PROMPT_H
+
+#include "libft.h"
+#include "directory.h"
```
Comment (s-t-a-n) [inline includes/prompt.h:5]: dont include directory.h in header, it is not part of api and it doesnt provide necessary datatypes

File: tests/src/prompt.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+#include "prompt.h"
+
+}
+
+ TEST_CASE( "Prompt basic testing", "[prompt]" ) {
+
+t_prompt *prompt;
+ prompt = prompt_create("user", "hostname");
+ prompt_print(prompt);
+     REQUIRE(1) ;
```
Comment (s-t-a-n) [inline tests/src/prompt.cpp:31]: this is a memory leak

call prompt_destroy()

File: tests/src/prompt.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+extern "C" {
+#include "prompt.h"
+
+}
+
+ TEST_CASE( "Prompt basic testing", "[prompt]" ) {
+
+t_prompt *prompt;
+ prompt = prompt_create("user", "hostname");
```
Comment (s-t-a-n) [inline tests/src/prompt.cpp:29]: verify result of prompt_create

-> REQUIRE(prompt != NULL);

File: tests/CMakeLists.txt [hunk 4]
```diff
@@ -47,3 +47,13 @@ target_link_libraries(${TEST}.bin PRIVATE Catch2::Catch2WithMain)
...
+
+## Prompt unittest
+set (TEST "test-prompt")
+set (SRC_LIST "src/prompt.cpp")
+add_executable(${TEST}.bin ${MAIN} ${SRC_LIST})
+target_compile_features(${TEST}.bin PRIVATE cxx_std_14)
+target_link_libraries(${TEST}.bin PRIVATE Catch2::Catch2WithMain)
+target_link_libraries(${TEST}.bin PRIVATE minishell ft vector hashmap)
+add_test(NAME ${TEST}
+		COMMAND ${TEST}.bin)
```
Comment (s-t-a-n) [inline tests/CMakeLists.txt:59]: perfect

File: src/prompt.c [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+	if (_prompt->username)
+		free(_prompt->username);
+	if (_prompt->hostname)
+		free(_prompt->hostname);
+	if (_prompt->current_dir)
+		free(_prompt->current_dir);
+	if (_prompt)
+		free(_prompt);
+}
```
Comment (s-t-a-n) [inline src/prompt.c:67]: you can leave the if tree
free does nothing when called with argument NULL

File: src/prompt.c [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+void	prompt_init_buffer(t_prompt *_prompt)
+{
+	ft_strlcat(_prompt->buffer, _prompt->username, ft_strlen(_prompt->username));
+	ft_strlcat(_prompt->buffer, "@", 1);
+	ft_strlcat(_prompt->buffer, _prompt->hostname, ft_strlen(_prompt->hostname));
+	ft_strlcat(_prompt->buffer, ": [", 1);
+	ft_strlcat(_prompt->buffer, _prompt->current_dir);
+	ft_strlcat(_prompt->buffer, "] %");
```
Comment (s-t-a-n) [inline src/prompt.c:32]: using this method, how do you know the index of where everything is in the buffer? 
use the indexes that point to the beginning of error_code, user, path, etc or use snprintf() to build the whole string

File: includes/prompt.h [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+t_prompt	*prompt_create(const char *_username, const char *_hostname);
+void		prompt_init_buffer(t_prompt *_prompt);
+void		prompt_update_path(t_prompt *_prompt);
+void		prompt_set_current_path(const char *_path);
+void 		prompt_set_error_code(int _error_code);
+void		prompt_print(t_prompt *_prompt);
+void		prompt_destroy(t_prompt *_prompt);
+
+#endif
```
Comment (s-t-a-n) [inline includes/prompt.h:27]: looks good!

**Review by s-t-a-n — CHANGES_REQUESTED** [review]
looks good man
just don't forget to use the indexes
don't be scared by pointer arithmetic, it's only doing ptr + int

and get that bloody submodule out of their

i think the least painfull way will be to make a new branch from develop
after gitpulling
and then do 'git checkout prompt [file name]' for each file to get the new versions, don't forget that if you do that on a Cmakelists,txt it could contain merges not present in your branch

**Pull request closed without merging**